### ISSUE TITLE
How to integrate Opengl ES to Navit

### ISSUE THREAD

**Issue by anantha-vidhya** [navit-gps/navit#1178]
Is there any way to modify navit source code so it uses opengl es not opengl since our embedded os supports opengl es package. I suspecting because of this the library is not linking properly which is showing in cmake step .

Thanks & Regards
A.N.V.Lavanya

**Comment by jkoan** [comment]
There is the graphics/egl plugin which should be compatible with OpenGL ES. If you would need to implement a graphics Plugin yourself here is the doc: https://wiki.navit-project.org/index.php/Graphic_driver_development 
Obviusly you need to configure navit to use graphics=egl which is described here: https://wiki.navit-project.org/index.php/Configuration/Display_Options#Graphics_Driver

**Comment by anantha-vidhya** [comment]
Thank you jokan.
 it is quiet interesting i will sure refer this and will get back to you.

**Comment by anantha-vidhya** [comment]
But here in 1st point in link you shared  it is mentioned that in navit.xml file we need to register the graphical driver. but navit.xml file itself is not there in source directory . it is created under build directory for me while doing for desktop but for arm board cmake itself failed.
Then how we will proceed. I had **navit.xml.python** under /home/futura/Desktop/Projects/NAVIT-Malasia-Demo/navit/navit/binding/python 
is it same what you refering to in source code directory.

**Comment by jkoan** [comment]
> but navit.xml file itself is not there in source directory . it is created under build directory for me while doing for desktop but for arm board cmake itself failed.

Yep this is right. It configures which plugins load wen navit is launched. The Python Binding has nothing to do with this.
Probably you need to fix the Navit build first. For your arm board you can probably disable all failing components which are not related to the config you intent to use. This should be done automatically, but its not always working.

**Comment by anantha-vidhya** [comment]
Hi Jkoan,

I try to follow your instruction, But here is the thing is navit.xml kind of file itself not created for me with navit source i cloned from git hub.
**git clone https://github.com/navit-gps/navit
mkdir navit-build
cd navit-build
cmake ../navit
make**

After cmake step the navit-build folder is filled with some builded folders, where navit.xml file is missing.

So in such what should i do where should i need to change the things?
Kindly please let me know.

Note: I downloaded navit0.5.3 directly from website in which navit.xml file is got created under /navit-build/navit/maps/ directory.
But this navit version is failing to open maps so why i didn't used this navit version.

Thanks & Regards
A.N.V.Lavanya

**Comment by anantha-vidhya** [comment]
Hi Jkoan,

When i am  trying to do cmake configuration the folder in build folder navit.xml file itself is not created with my cross-compile environment.

Thanks & Regards
A.N.V.Lavanya

**Comment by jkoan** [comment]
Could you please provide the console output of everything you did to try to build navit? Please upload at gist.github.com and share the link.
At the bare minimum I need the full cmake output and the build output (normally make)

**Comment by anantha-vidhya** [comment]
https://gist.github.com/anantha-vidhya/f2e55c67c1a04ed779e51a2694ab8372

**Comment by anantha-vidhya** [comment]
1.I actually fount the graphics files are the one which is responsible for graphics rendering on screen.So i am trying to replace all the functions with compatible opengl es functions so that graphics file will get modified to opengl es compatible. Then i will change navit.c file so that application files will change according to library compatible 
All the way i am thinking to proceed is it correct way? 
Correct me if my attempt of doing is wrong.
Thanks & Regards
A.N.V.Lavanya

**Comment by anantha-vidhya** [comment]
Please some one respond to me. Any sort of help is appreciated.

**Comment by jkoan** [comment]
> https://gist.github.com/anantha-vidhya/f2e55c67c1a04ed779e51a2694ab8372

Please try to disable the modules which are failing. `cmake -L` gives you all the names for that. For the currently failing module its `-Dgraphics/qt_qpainter=FALSE`



> 1.I actually fount the graphics files are the one which is responsible for graphics rendering on screen.So i am trying to replace all the functions with compatible opengl es functions so that graphics file will get modified to opengl es compatible. Then i will change navit.c file so that application files will change according to library compatible All the way i am thinking to proceed is it correct way? Correct me if my attempt of doing is wrong. Thanks & Regards A.N.V.Lavanya

I don't know if you already understood navits modular build approach. under navit/graphics you see multiple plugins which can be build. Some of which are more desktop oriented, others that work very well on embedded. So you need to look into which specific one you wane use. As you said you need OpenGL ES so i would amuse that the "opengl" Plugin would be feasible or the egl plugin. I don't currently know if one of them uses functions which are unsupported on opengl es. But what i can say is that you only need one graphics plugin. 



> Please some one respond to me. Any sort of help is appreciated.

I want to stress that you don't paying anyone to do work for you. So please don't stress someone to take time for you in their free time. And try to proactively provide all the information you can like how to reproduce your exact problem. For example, i think you are using a build open source build environment. Why don't you provide every step necessary to reproduce you problem?

**Comment by anantha-vidhya** [comment]
``I want to stress that you don't paying anyone to do work for you. So please don't stress someone to take time for you in their free time. ``

Sorry for framing out sentence in such way but i mean it as request to gain the knowledge.
I know that i am not paying for you anything,So please apology me if i mean in such  a way.
So far for all the help Thanking you.

``And try to proactively provide all the information you can like how to reproduce your exact problem. For example, i think you are using a build open source build environment. Why don't you provide every step necessary to reproduce you problem?``
Yeah Sure.

1. Firstly i downloaded the ti-processor-sdk-linux-am437x-evm-04.02.00.09-Linux-x86-Install.bin from the below link,
https://www.ti.com/tool/download/PROCESSOR-SDK-LINUX-AM437X/04.02.00.09
2. I did extracted the file by running ./ti-processor-sdk-linux-am437x-evm-04.02.00.09-Linux-x86-Install.bin
3. It is the SDK folder with all the tool chain environment and all necessary file system things containing.
4. Sourced the Cross-compiling tool chain with the following step.
          cd /home/<username>/ti-processor-sdk-linux-am437x-evm-04.02.00.09/linux-devkit/
          source **environment-setup**
5. The terminal prompt will change as **[linux-devkit]:~>** 
6. Then
     mkdir navit-build-arm
     cd /navit-build-arm
     cmake ../navit
     make 

7. getting errors at make step.
8. Before step 4 i did one more step i.e i builded some SDL-Image library, ttf-liberation, etc dependencies libraries using bit bake method dedicated to this SDK and those packages i copied to the **linux-devkit/sysroots/armv7ahf-neon-linux-gnueabi** as well as **linux-devkit/sysroots/x86_64-arago-linux**
9. After that i in cmake step we can see SDL found.. kind of messages but for opengl i couldn't able to build the proper package so i couldn't able to copy the builded package.
10. How i builded the packages for arm  using bitbake method.
      - sudo dpkg-reconfigure dash
      # Select "No" when prompted
      - wget https://releases.linaro.org/components/toolchain/binaries/6.2-2016.11/arm-linux-gnueabihf/gcc-linaro-6.2.1-2016.11-x86_64_arm-linux-gnueabihf.tar.xz
      - tar -Jxvf gcc-linaro-6.2.1-2016.11-x86_64_arm-linux-gnueabihf.tar.xz -C $HOME
      - git clone git://arago-project.org/git/projects/oe-layersetup.git tisdk
      - cd tisdk
      - ./oe-layertool-setup.sh -f configs/processor-sdk/processor-sdk-04.02.00.09-config.txt
      -  cd build
      - . conf/setenv
      - export PATH=$HOME/gcc-linaro-6.2.1-2016.11-x86_64_arm-linux-gnueabihf/bin:$PATH
      - MACHINE=am437x-evm bitbake <pkg_name>
      - So by this method we will be builded pkg folder under 
         /tisdk/build/arago-tmp-external-linaro-toolchain/work/armv7ahf-neon-linux-gnueabi/pkg_name

11. Under these folders choose deploy-ipks folder for target and image folder for host(to copy library .so files to linux-devkit as mentioned in step8)

Hope this all the information will help you in exactly describing my issue.Else if you require more information, i am ready to give.

Thanks & Regards
A.N.V.Lavanya
